netbsd/usr.sbin/lpr/lpd/lpd.8
.\"	$NetBSD: lpd.8,v 1.17.4.1 2001/04/26 12:49:12 he Exp $
.\"
.\" Copyright (c) 1983, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"	This product includes software developed by the University of
.\"	California, Berkeley and its contributors.
.\" 4. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     @(#)lpd.8	8.3 (Berkeley) 4/19/94
.\"
.Dd April 19, 1994
.Dt LPD 8
.Os
.Sh NAME
.Nm lpd
.Nd line printer spooler daemon
.Sh SYNOPSIS
.Nm ""
.Op Fl dlsr
.Op Fl n Ar maxchild
.Op Fl w Ar maxwait
.Op port
.Sh DESCRIPTION
.Nm
is the line printer daemon (spool area handler) and is normally invoked
at boot time from the
.Xr rc 8
file.
It makes a single pass through the
.Xr printcap 5
file to find out about the existing printers and prints any files
left after a crash.
It then uses the system calls
.Xr listen 2
and
.Xr accept 2
to receive requests to print files in the queue, transfer files to
the spooling area, display the queue, or remove jobs from the queue.
In each case, it forks a child to handle the request so the parent
can continue to listen for more requests.
.Pp
Available options:
.Bl -tag -width Ds
.It Fl d
The
.Fl d
option turns on the
.Dv SO_DEBUG
.Xr socket 2
option.  See
.Xr setsockopt 2
for more details.
.It Fl l
The
.Fl l
flag causes
.Nm
to log valid requests received from the network.
This can be useful for debugging purposes.
.It Fl n
The
.Fl n
flag sets
.Ar maxchild
as the maximum number of child processes that
.Nm
will spawn.  The default is 32.
.It Fl r
The
.Fl r
flag allows the
.Dq of
filter to be use if specified for a remote
printer.  Traditionally,
.Nm
would not use the output filter for remote printers.
.It Fl s
The
.Fl s
flag selects
.Dq secure
mode, in which
.Nm
does not listen on a TCP socket but only takes commands from a
.Ux
domain socket.
This is valuable when the machine on which
.Nm
runs is subject to attack over the network and it is desired that the
machine be protected from attempts to remotely fill spools and similar
attacks.
.It Fl w
The
.Fl w
flag sets
.Ar maxwait
as the wait time (in seconds) for dead remote server detection.  If
no response is returned from a connected server within this period,
the connection is closed and a message logged.  The default is
120 seconds.
.El
.Pp
If the
.Op port
parameter is passed,
.Nm
listens on this port instead of the usual
.Dq printer/tcp
port from
.Pa /etc/services .
.Pp
Access control is provided by two means.
First, all requests must come from one of the machines listed in
the file
.Pa /etc/hosts.equiv
or
.Pa /etc/hosts.lpd .
Second, if the
.Li rs
capability is specified in the
.Xr printcap 5
entry for the printer being accessed,
.Em lpr
requests will only be honored for those users with accounts on the
machine with the printer.
.Pp
The file
.Em minfree
in each spool directory contains the number of disk blocks to leave free
so that the line printer queue won't completely fill the disk.
The
.Em minfree
file can be edited with your favorite text editor.
.Pp
The daemon begins processing files
after it has successfully set the lock for exclusive
access (described a bit later),
and scans the spool directory
for files beginning with
.Em cf .
Lines in each
.Em cf
file specify files to be printed or non-printing actions to be performed.
Each such line begins with a key character to specify what to do
with the remainder of the line.
.Bl -tag -width Ds
.It J
Job Name.
String to be used for the job name on the burst page.
.It C
Classification.
String to be used for the classification line
on the burst page.
.It L
Literal.
The line contains identification info from the password file and
causes the banner page to be printed.
.It T
Title.
String to be used as the title for
.Xr pr 1 .
.It H
Host Name.
Name of the machine where
.Xr lpr 1
was invoked.
.It P
Person.
Login name of the person who invoked
.Xr lpr 1 .
This is used to verify ownership by
.Xr lprm 1 .
.It M
Send mail to the specified user when the current print job completes.
.It f
Formatted File.
Name of a file to print which is already formatted.
.It l
Like
.Dq f
but passes control characters and does not make page breaks.
.It p
Name of a file to print using
.Xr pr 1
as a filter.
.It t
Troff File.
The file contains
.Xr troff 1
output (cat phototypesetter commands).
.It n
Ditroff File.
The file contains device independent troff
output.
.It r
DVI File.
The file contains
.Tn Tex l
output
DVI format from Standford.
.It g
Graph File.
The file contains data produced by
.Xr plot 3 .
.It c
Cifplot File.
The file contains data produced by
.Em cifplot .
.It v
The file contains a raster image.
.It r
The file contains text data with
FORTRAN carriage control characters.
.It \&1
Troff Font R.
Name of the font file to use instead of the default.
.It \&2
Troff Font I.
Name of the font file to use instead of the default.
.It \&3
Troff Font B.
Name of the font file to use instead of the default.
.It \&4
Troff Font S.
Name of the font file to use instead of the default.
.It W
Width.
Changes the page width (in characters) used by
.Xr pr 1
and the text filters.
.It I
Indent.
The number of characters to indent the output by (in ascii).
.It U
Unlink.
Name of file to remove upon completion of printing.
.It N
File name.
The name of the file which is being printed, or a blank for the
standard input (when
.Xr lpr 1
is invoked in a pipeline).
.El
.Pp
If a file cannot be opened, a message will be logged via
.Xr syslog 3
using the
.Em LOG_LPR
facility.
.Nm
will try up to 20 times to reopen a file it expects to be there,
after which it will skip the file to be printed.
.Pp
.Nm
uses
.Xr flock 2
to provide exclusive access to the lock file and to prevent multiple
daemons from becoming active simultaneously.
If the daemon should be killed or die unexpectedly, the lock file
need not be removed.
The lock file is kept in a readable
.Tn ASCII
form and contains two lines.
The first is the process id of the daemon and the second is the control
file name of the current job being printed.
The second line is updated to reflect the current status of
.Nm
for the programs
.Xr lpq 1
and
.Xr lprm 1 .
.Sh FILES
.Bl -tag -width "/var/spool/output/*/minfree" -compact
.It Pa /etc/printcap
printer description file
.It Pa /var/spool/output/*
spool directories
.It Pa /var/spool/output/*/minfree
minimum free space to leave
.It Pa /dev/lp*
line printer devices
.It Pa /var/run/printer
socket for local requests
.It Pa /etc/hosts.equiv
lists machine names allowed printer access
.It Pa /etc/hosts.lpd
lists machine names allowed printer access,
but not under same administrative control.
.El
.Sh SEE ALSO
.Xr lpq 1 ,
.Xr lpr 1 ,
.Xr lprm 1 ,
.Xr setsockopt 2 ,
.Xr syslog 3 ,
.Xr hosts.equiv 5 ,
.Xr printcap 5 ,
.Xr lpc 8 ,
.Xr pac 8
.Rs
.%T "4.3 BSD Line Printer Spooler Manual"
.Re
.Sh HISTORY
An
.Nm

CHANGELOG
$KAME: CHANGELOG,v 1.2458 2003/08/18 05:39:09 itojun Exp $

<200308>
Sun Aug 17 JST 2003  itojun@iijlab.net
	* sys/netinet{,6}/in{,6}_proto.c: enforce inbound IPsec policy on
	  protocols that falls into raw socket as a wildcard case.

2003-08-15  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/kame/pim6sd/{cfparse.y, mrt.[ch], pim6_proto.c, rp.c}
	Now static-RP configuration can work together with bootstrap-
	message-based RP configuration

2003-08-10  SUZUKI, Shinsuke <suz@crl.hitachi.co.jp>
	* kame/{freebsd5, kame}
	sync with FreeBSD-5.1 RELEASE (behaviour is not confirmed yet)

2003-08-07  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* kame/kame/had/mpa.c,haadisc.c
	A home address should be passed in the source address of a mobile
	solicit pakcet. This problem was reported and donated the code by
	Kenichi Yajima <yajima@netlab.nec.co.jp>
	
2003-08-07  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/cftoken.l: use a stronger lexical check for
	interface names.
	From: Hajimu UMEMOTO <ume@mahoroba.org>
	* kame/kame/dhcp6/config.c (configure_interface): check interfaces
	specified in the configuration file really exist.

2003-08-05  Hideki ONO  <ono@soft.net.fujitsu.co.jp>
	* removed vrrp6 support because it might be a patent infringement.
	  http://www.ietf.org/ietf/IPR/cisco-ipr-draft-ietf-vrrp-ipv6-spec.txt

2003-08-01  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6: implemented various clarifications and some new
	  features:
	- supported DNS name and NTP servers options
	- dhcp6s now uses the server/relay port to send a relay-rely
	  message
	- dhcp6s now ignores a rebind message when it cannot find a
	  binding, according to prefix-delegation-04
	- dhcp6relay now accepts relay-reply messages on the socket
	  listening on the server/relay port
	* The DHCPv6 implementation is now built by default under
	  *bsd/usr.sbin/.

<200307>
Thu Jul 31 19:19:48 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.c,mip6_cncore.c
	sending/receiving a binding refresh request message is supported.

Tue Jul 29 17:09:15 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_mncore.[hc]
	select a CoA using the source address selection like algorithm.
	suggested by Francis.Dupont@enst-bretagne.fr.

2003-07-28  MOMOSE Tsuyoshi  <t-momose@netlab.nec.co.jp>
	* sys/netinet6/{ip6_output.c, ip6_var.h, mip6_cncore.c,
	mip6_hacore.[ch], nd6.c}:
	CN or HA dosen't add a type2 routing header to a Binding
	Acknowledgement packet when a deregistration packet was issued
	on home link and it was failed by some reasons.
	reported by Yukiyo.Akisada@jp.yokogawa.com and v6pc cert.
	team.

Mon Jul 28 20:03:46 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_cncore.c,mip6_mncore.[hc],in6.c,nd6_rtr.c
	  kame/sys/net/if_hif.[hc]
	assign a CoA separately for each home interface, since a home
	interface is logically independent from each other.

Mon Jul 28 14:38:11 JST 2003	keiichi@iij.ad.jp
	* kame/sys/netinet6/mip6_hacore.c,mip6_cncore.c:
	- fixed a bug not sending a binding ack when DAD for a link-local
	  address is failed.  (en-bugged during re-structuring BC logic.)
	- fixed a bug using RTHDR2 with Mobility Headers other than
	  a binding ack.
	reported by Yukiyo.Akisada@jp.yokogawa.com.

Fri Jul 25 19:12:15 JST 2003  itojun@iijlab.net
	* sys/netinet/ah_core.c: hmac-ripemd160 support

Fri Jul 25 18:07:43 JST 2003 sakane@kame.net
	* kame/sys/netkey/key.c
	fixed that the kernel crashed when key_spdacquire() was called
	because key_spdacquire() had been implemented imcopletely.

Thu Jul 24 16:11:06 JST 2003	keiichi@iij.ad.jp
	* mip6 related files.
	simplify the prefix and advertising router list management
	mechanism on a mobile node.

2003-07-20  JINMEI, Tatuya  <jinmei@isl.rdc.toshiba.co.jp>
	* kame/kame/dhcp6/config.c (configure_duid): fixed a bug of
	allocating short memory.  Upgrading is required if you configure
	dhcp6s with the duid statement.

Sat Jul 19 18:12:01 JST 2003  itojun@iijlab.net
	* sys/netinet6/esp_aesctr.c: support draft-ietf-ipsec-ciph-aes-ctr-03
	  as there's no official DOI assignment yet, we use private DOI number.
	* sys/netinet6/ah_aesxcbcmac.c: support
	  draft-ietf-ipsec-ciph-aes-xcbc-mac-03.

freebsd2/usr.sbin/lpr/lpd/lpd.c
/*
 * Copyright (c) 1983, 1993, 1994
 *	The Regents of the University of California.  All rights reserved.
 *
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *	This product includes software developed by the University of
 *	California, Berkeley and its contributors.
 * 4. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 */

#ifndef lint
static const char copyright[] =
"@(#) Copyright (c) 1983, 1993, 1994\n\
	The Regents of the University of California.  All rights reserved.\n";
#endif /* not lint */

#ifndef lint
#if 0
static char sccsid[] = "@(#)lpd.c	8.7 (Berkeley) 5/10/95";
#endif
static const char rcsid[] =
	"$Id: lpd.c,v 1.3.2.5 1997/09/25 06:32:39 charnier Exp $";
#endif /* not lint */

/*
 * lpd -- line printer daemon.
 *
 * Listen for a connection and perform the requested operation.
 * Operations are:
 *	\1printer\n
 *		check the queue for jobs and print any found.
 *	\2printer\n
 *		receive a job from another machine and queue it.
 *	\3printer [users ...] [jobs ...]\n
 *		return the current state of the queue (short form).
 *	\4printer [users ...] [jobs ...]\n
 *		return the current state of the queue (long form).
 *	\5printer person [users ...] [jobs ...]\n
 *		remove jobs from the queue.
 *
 * Strategy to maintain protected spooling area:
 *	1. Spooling area is writable only by daemon and spooling group
 *	2. lpr runs setuid root and setgrp spooling group; it uses
 *	   root to access any file it wants (verifying things before
 *	   with an access call) and group id to know how it should
 *	   set up ownership of files in the spooling area.
 *	3. Files in spooling area are owned by root, group spooling
 *	   group, with mode 660.
 *	4. lpd, lpq and lprm run setuid daemon and setgrp spooling group to
 *	   access files and printer.  Users can't get to anything
 *	   w/o help of lpq and lprm programs.
 */

#include <sys/param.h>
#include <sys/wait.h>
#include <sys/types.h>
#include <sys/socket.h>
#include <sys/un.h>
#include <sys/stat.h>
#include <sys/file.h>
#include <netinet/in.h>
#include <arpa/inet.h>

#include <netdb.h>
#include <unistd.h>
#include <syslog.h>
#include <signal.h>
#include <err.h>
#include <errno.h>
#include <fcntl.h>
#include <dirent.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sysexits.h>
#include <ctype.h>
#include "lp.h"
#include "lp.local.h"
#include "pathnames.h"
#include "extern.h"

int	lflag;				/* log requests flag */
int	from_remote;			/* from remote socket */

static void       reapchild __P((int));
static void       mcleanup __P((int));
static void       doit __P((void));
static void       startup __P((void));
static void       chkhost __P((struct sockaddr *));
static int	  ckqueue __P((char *));
static void	  usage __P((void));

uid_t	uid, euid;

u_char family = AF_INET;

int
main(argc, argv)
	int argc;
	char **argv;
{
	int f, funix, finet, finet6, options, fromlen, i, errs;
	fd_set defreadfds;
	struct sockaddr_un un, fromunix;
	struct sockaddr_in sin, frominet;
	struct sockaddr_in6 sin6, frominet6;
	int omask, lfd;
	struct servent *sp, serv;
	int inet_flag = 0, inet6_flag = 0;

	euid = geteuid();	/* these shouldn't be different */
	uid = getuid();
	options = 0;
	gethostname(host, sizeof(host));

	name = "lpd";

	if (euid != 0)
		errx(EX_NOPERM,"must run as root");

	errs = 0;
	while ((i = getopt(argc, argv, "dl46")) != -1)
		switch (i) {
		case 'd':
			options |= SO_DEBUG;
			break;
		case 'l':
			lflag++;
			break;
		case '4':
			family = AF_INET;
			inet_flag++;
			break;
		case '6':
			family = AF_INET6;
			inet6_flag++;
			break;
		default:
			errs++;
		}
	if (inet_flag && inet6_flag)
		family = AF_UNSPEC;
	argc -= optind;
	argv += optind;
	if (errs)
		usage();

	if (argc == 1) {
		if ((i = atoi(argv[0])) == 0)
			usage();
		if (i < 0 || i > USHRT_MAX)
			errx(EX_USAGE, "port # %d is invalid", i);

		serv.s_port = htons(i);
		sp = &serv;
		argc--;
	} else {
		sp = getservbyname("printer", "tcp");
		if (sp == NULL)
			errx(EX_OSFILE, "printer/tcp: unknown service");
	}

	if (argc != 0)
		usage();

#ifndef DEBUG
	/*
	 * Set up standard environment by detaching from the parent.
	 */
	daemon(0, 0);
#endif

	openlog("lpd", LOG_PID, LOG_LPR);
	syslog(LOG_INFO, "restarted");
	(void) umask(0);
	lfd = open(_PATH_MASTERLOCK, O_WRONLY|O_CREAT, 0644);
	if (lfd < 0) {
		syslog(LOG_ERR, "%s: %m", _PATH_MASTERLOCK);
		exit(1);
	}
	if (flock(lfd, LOCK_EX|LOCK_NB) < 0) {
		if (errno == EWOULDBLOCK)	/* active deamon present */
			exit(0);
		syslog(LOG_ERR, "%s: %m", _PATH_MASTERLOCK);
		exit(1);
	}
	ftruncate(lfd, 0);
	/*
	 * write process id for others to know
	 */
	sprintf(line, "%u\n", getpid());
	f = strlen(line);
	if (write(lfd, line, f) != f) {
		syslog(LOG_ERR, "%s: %m", _PATH_MASTERLOCK);
		exit(1);
	}
	signal(SIGCHLD, reapchild);
	/*
	 * Restart all the printers.
	 */
	startup();
	(void) unlink(_PATH_SOCKETNAME);
	funix = socket(AF_UNIX, SOCK_STREAM, 0);
	if (funix < 0) {
		syslog(LOG_ERR, "socket: %m");
		exit(1);
	}
#define	mask(s)	(1 << ((s) - 1))
	omask = sigblock(mask(SIGHUP)|mask(SIGINT)|mask(SIGQUIT)|mask(SIGTERM));
	(void) umask(07);
	signal(SIGHUP, mcleanup);
	signal(SIGINT, mcleanup);
	signal(SIGQUIT, mcleanup);
	signal(SIGTERM, mcleanup);
	memset(&un, 0, sizeof(un));
	un.sun_family = AF_UNIX;
	strcpy(un.sun_path, _PATH_SOCKETNAME);
#ifndef SUN_LEN
#define SUN_LEN(unp) (strlen((unp)->sun_path) + 2)
#endif
	if (bind(funix, (struct sockaddr *)&un, SUN_LEN(&un)) < 0) {
		syslog(LOG_ERR, "ubind: %m");
		exit(1);
	}
	(void) umask(0);
	sigsetmask(omask);
	FD_ZERO(&defreadfds);
	FD_SET(funix, &defreadfds);
	listen(funix, 5);
    if (family == AF_INET || family == AF_UNSPEC) {
	finet = socket(AF_INET, SOCK_STREAM, 0);
	if (finet >= 0) {
		if (options & SO_DEBUG)
			if (setsockopt(finet, SOL_SOCKET, SO_DEBUG, 0, 0) < 0) {
				syslog(LOG_ERR, "setsockopt (SO_DEBUG): %m");
				mcleanup(0);
			}
		memset(&sin, 0, sizeof(sin));
		sin.sin_family = AF_INET;
		sin.sin_port = sp->s_port;
		if (bind(finet, (struct sockaddr *)&sin, sizeof(sin)) < 0) {
			syslog(LOG_ERR, "bind: %m");
			mcleanup(0);
		}
		FD_SET(finet, &defreadfds);
		listen(finet, 5);
	}
    }
    if (family == AF_INET6 || family == AF_UNSPEC) {
	finet6 = socket(AF_INET6, SOCK_STREAM, 0);
	if (finet6 >= 0) {
		if (options & SO_DEBUG)
			if (setsockopt(finet6, SOL_SOCKET, SO_DEBUG, 0, 0) < 0) {
				syslog(LOG_ERR, "setsockopt (SO_DEBUG): %m");
				mcleanup(0);
			}
		memset(&sin6, 0, sizeof(sin6));
		sin6.sin6_family = AF_INET6;
		sin6.sin6_port = sp->s_port;
		if (bind(finet6, (struct sockaddr *)&sin6, sizeof(sin6)) < 0) {
			syslog(LOG_ERR, "bind: %m");
			mcleanup(0);
		}
		FD_SET(finet6, &defreadfds);
		listen(finet6, 5);
	}
    }
	/*
	 * Main loop: accept, do a request, continue.
	 */
	memset(&frominet, 0, sizeof(frominet));
	memset(&frominet6, 0, sizeof(frominet6));
	memset(&fromunix, 0, sizeof(fromunix));
	for (;;) {
		int domain, nfds, s;
		fd_set readfds;

		FD_COPY(&defreadfds, &readfds);
		nfds = select(20, &readfds, 0, 0, 0);
		if (nfds <= 0) {
			if (nfds < 0 && errno != EINTR)
				syslog(LOG_WARNING, "select: %m");
			continue;
		}
		if (FD_ISSET(funix, &readfds)) {
			domain = AF_UNIX, fromlen = sizeof(fromunix);
			s = accept(funix,
			    (struct sockaddr *)&fromunix, &fromlen);
		} else if (finet6 >= 0 && FD_ISSET(finet6, &readfds))  {
			domain = AF_INET6, fromlen = sizeof(frominet6);
			s = accept(finet6,
			    (struct sockaddr *)&frominet6, &fromlen);
			if (frominet6.sin6_port == htons(20)) {
				close(s);
				continue;
			}
		} else if (finet >= 0 && FD_ISSET(finet, &readfds)) {
			domain = AF_INET, fromlen = sizeof(frominet);
			s = accept(finet,
			    (struct sockaddr *)&frominet, &fromlen);
			if (frominet.sin_port == htons(20)) {
				close(s);
				continue;
			}
		}
		if (s < 0) {
			if (errno != EINTR)
				syslog(LOG_WARNING, "accept: %m");
			continue;
		}
		if (fork() == 0) {
			signal(SIGCHLD, SIG_IGN);
			signal(SIGHUP, SIG_IGN);
			signal(SIGINT, SIG_IGN);
			signal(SIGQUIT, SIG_IGN);
			signal(SIGTERM, SIG_IGN);
			(void) close(funix);
			if (finet >= 0)
				(void) close(finet);
			if (finet6 >= 0)
				(void) close(finet6);
			dup2(s, 1);
			(void) close(s);
			if (domain == AF_INET || domain == AF_INET6) {
				from_remote = 1;
				chkhost((domain == AF_INET ?
					(struct sockaddr *) &frominet :
					(struct sockaddr *) &frominet6));
			} else
				from_remote = 0;
			doit();
			exit(0);
		}
		(void) close(s);
	}
}

static void
reapchild(signo)
	int signo;
{
	union wait status;

	while (wait3((int *)&status, WNOHANG, 0) > 0)
		;
}

static void
mcleanup(signo)
	int signo;
{
	if (lflag)
		syslog(LOG_INFO, "exiting");
	unlink(_PATH_SOCKETNAME);
	exit(0);
}

/*
 * Stuff for handling job specifications
 */
char	*user[MAXUSERS];	/* users to process */
int	users;			/* # of users in user array */
int	requ[MAXREQUESTS];	/* job number of spool entries */
int	requests;		/* # of spool requests */
char	*person;		/* name of person doing lprm */

char	fromb[MAXHOSTNAMELEN];	/* buffer for client's machine name */
char	cbuf[BUFSIZ];		/* command line buffer */
char	*cmdnames[] = {
	"null",
	"printjob",
	"recvjob",
	"displayq short",
	"displayq long",
	"rmjob"
};

static void
doit()
{
	register char *cp;
	register int n;

	for (;;) {
		cp = cbuf;
		do {
			if (cp >= &cbuf[sizeof(cbuf) - 1])
				fatal("Command line too long");
			if ((n = read(1, cp, 1)) != 1) {
				if (n < 0)
					fatal("Lost connection");
				return;
			}
		} while (*cp++ != '\n');
		*--cp = '\0';
		cp = cbuf;
		if (lflag) {
			if (*cp >= '\1' && *cp <= '\5')
				syslog(LOG_INFO, "%s requests %s %s",
					from, cmdnames[*cp], cp+1);
			else
				syslog(LOG_INFO, "bad request (%d) from %s",
					*cp, from);
		}
		switch (*cp++) {
		case '\1':	/* check the queue and print any jobs there */
			printer = cp;
			printjob();
			break;
		case '\2':	/* receive files to be queued */
			if (!from_remote) {
				syslog(LOG_INFO, "illegal request (%d)", *cp);
				exit(1);
			}
			printer = cp;
			recvjob();
			break;
		case '\3':	/* display the queue (short form) */
		case '\4':	/* display the queue (long form) */
			printer = cp;
			while (*cp) {
				if (*cp != ' ') {
					cp++;
					continue;
				}
				*cp++ = '\0';
				while (isspace(*cp))
					cp++;
				if (*cp == '\0')
					break;
				if (isdigit(*cp)) {
					if (requests >= MAXREQUESTS)
						fatal("Too many requests");
					requ[requests++] = atoi(cp);
				} else {
					if (users >= MAXUSERS)
						fatal("Too many users");
					user[users++] = cp;
				}
			}
			displayq(cbuf[0] - '\3');
			exit(0);
		case '\5':	/* remove a job from the queue */
			if (!from_remote) {
				syslog(LOG_INFO, "illegal request (%d)", *cp);
				exit(1);
			}
			printer = cp;
			while (*cp && *cp != ' ')
				cp++;
			if (!*cp)
				break;
			*cp++ = '\0';
			person = cp;
			while (*cp) {
				if (*cp != ' ') {
					cp++;
					continue;
				}
				*cp++ = '\0';
				while (isspace(*cp))
					cp++;
				if (*cp == '\0')
					break;
				if (isdigit(*cp)) {
					if (requests >= MAXREQUESTS)
						fatal("Too many requests");
					requ[requests++] = atoi(cp);
				} else {
					if (users >= MAXUSERS)
						fatal("Too many users");
					user[users++] = cp;
				}
			}
			rmjob();
			break;
		}
		fatal("Illegal service request");
	}
}

/*
 * Make a pass through the printcap database and start printing any
 * files left from the last time the machine went down.
 */
static void
startup()
{
	char *buf;
	register char *cp;
	int pid;
	char *spooldirs[16];        /* Which spooldirs are active? */
	int i;                      /* Printer index presently processed */
	int j;                      /* Printer index of potential conflict */
	char *spooldir;             /* Spooldir of present printer */
	int  canfreespool;          /* Is the spooldir malloc()ed? */

	/*
	 * Restart the daemons and test for spooldir conflict.
	 */
	i = 0;
	while (cgetnext(&buf, printcapdb) > 0) {

		/* Check for duplicate spooldirs */
		canfreespool = 1;
		if (cgetstr(buf, "sd", &spooldir) <= 0) {
			spooldir = _PATH_DEFSPOOL;
			canfreespool = 0;
		}
		if (i < sizeof(spooldirs)/sizeof(spooldirs[0]))
			spooldirs[i] = spooldir;
		for (j = 0;
			 j < MIN(i,sizeof(spooldirs)/sizeof(spooldirs[0]));
			 j++) {
			if (strcmp(spooldir, spooldirs[j]) == 0) {
				syslog(LOG_ERR,
					"startup: duplicate spool directories: %s",
					spooldir);
				mcleanup(0);
			}
		}
		if (canfreespool && i >= sizeof(spooldirs)/sizeof(spooldirs[0]))
			free(spooldir);
		i++;
		/* Spooldir test done */

		if (ckqueue(buf) <= 0) {
			free(buf);
			continue;	/* no work to do for this printer */
		}
		for (cp = buf; *cp; cp++)
			if (*cp == '|' || *cp == ':') {
				*cp = '\0';
				break;
			}
		if (lflag)
			syslog(LOG_INFO, "work for %s", buf);
		if ((pid = fork()) < 0) {
			syslog(LOG_WARNING, "startup: cannot fork");
			mcleanup(0);
		}
		if (!pid) {
			printer = buf;
			cgetclose();
			printjob();
			/* NOTREACHED */
		}
		else free(buf);
	}
}

/*
 * Make sure there's some work to do before forking off a child
 */
static int
ckqueue(cap)
	char *cap;
{
	register struct dirent *d;
	DIR *dirp;
	char *spooldir;

	if (cgetstr(cap, "sd", &spooldir) == -1)
		spooldir = _PATH_DEFSPOOL;
	if ((dirp = opendir(spooldir)) == NULL)
		return (-1);
	while ((d = readdir(dirp)) != NULL) {
		if (d->d_name[0] != 'c' || d->d_name[1] != 'f')
			continue;	/* daemon control files only */
		closedir(dirp);
		return (1);		/* found something */
	}
	closedir(dirp);
	return (0);
}

#define DUMMY ":nobody::"

/*
 * Check to see if the from host has access to the line printer.
 */
static void
chkhost(f)
	struct sockaddr *f;
{
	register struct hostent *hp;
	register FILE *hostf;
	int first = 1;
	int good = 0;
	int err;
	char numerichost[INET6_ADDRSTRLEN];
	caddr_t addr = f->sa_family == AF_INET ?
			(caddr_t) &((struct sockaddr_in *) f)->sin_addr :
			(caddr_t) &((struct sockaddr_in6 *) f)->sin6_addr;

	/* Need real hostname for temporary filenames */
	err = getnameinfo(f, f->sa_len, numerichost, sizeof(numerichost), 0, 0, NI_NUMERICHOST);
	if (err) {
		syslog(LOG_ERR, "getnameinfo: failed %d", err);
		exit(1);
	}
	err = getnameinfo(f, f->sa_len, fromb, sizeof(fromb), 0, 0, NI_NAMEREQD);
	if (err)
		fatal("Host name for your address (%s) unknown", numerichost);
	from = fromb;

	/* Check for spoof, ala rlogind */
	hp = gethostbyname2(fromb, f->sa_family);
	if (!hp)
		fatal("hostname for your address (%s) unknown", numerichost);
	for (; good == 0 && hp->h_addr_list[0] != NULL; hp->h_addr_list++) {
		if (!bcmp(hp->h_addr_list[0], addr, hp->h_length))
			good = 1;
	}
	if (good == 0) {
		fatal("address for your hostname (%s) not matched", numerichost);
	}

	hostf = fopen(_PATH_HOSTSEQUIV, "r");
again:
	if (hostf) {
		if (__ivaliduser_af(hostf, addr,
		    DUMMY, DUMMY,
		    hp->h_addrtype, hp->h_length) == 0) {
			(void) fclose(hostf);
			return;
		}
		(void) fclose(hostf);
	}
	if (first == 1) {
		first = 0;
		hostf = fopen(_PATH_HOSTSLPD, "r");
		goto again;
	}
	fatal("Your host does not have line printer access");
	/*NOTREACHED*/
}

static void
usage()
{
	fprintf(stderr, "usage: lpd [-dl] [port#]\n");
	exit(EX_USAGE);
}
